crates/pessimistic-proof-utils/src/smt.rs

```diff
@@ -0,0 +1,559 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use pessimistic_proof::local_exit_tree::hasher::Hasher;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use thiserror::Error;
+
+use crate::utils::empty_hash_at_height;
+
+pub trait ToBits<const NUM_BITS: usize> {
+    fn to_bits(&self) -> [bool; NUM_BITS];
+}
+
+#[derive(Error, Debug, Eq, PartialEq)]
+pub(crate) enum SmtError {
+    #[error("trying to insert a key already in the SMT")]
+    KeyAlreadyPresent,
+    #[error("trying to generate a Merkle proof for a key not in the SMT")]
+    KeyNotPresent,
+    #[error("trying to generate a non-inclusion proof for a key present in the SMT")]
+    KeyPresent,
+}
+
+/// A node in an SMT.
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Node<H>
+where
+    H: Hasher,
+    H::Digest: Serialize + DeserializeOwned,
```

Comment:
So I agree in principle, however here I think it would negatively impact the code's readability and succinctness to remove this (because we'd need to add a custom impl of `(De)Serialize` when `H::Digest: Serialize + DeserializeOwned` instead of a clean derive macro).
Also, I don't think we'd ever want to work with non-serializable hashes, so that's a way to enforce it in the code.

But tbh I don't think we actually need these structs (especially `Smt`) to be (de)serializable, so at some point we could just remove these derive macros altogether.

Reply:
> I think it would negatively impact the code's readability

Agreed entirely